File: go/rewind_test.go
package tabnas

// Token-rewind primitives on Context: Mark() captures the current parse
// position; Rewind(m) replays every token consumed since that mark by
// re-feeding the active lexer's pending-token queue. Mirrors
// ts/test/rewind.test.js. Go reports an out-of-window rewind as an error
// (TS throws) to preserve the no-panic guarantee.

import (
	"strings"
	"testing"
)

// rewindParser builds a bare-engine instance with the given fixed tokens
// and rewind options, then lets the caller install rules.
func rewindParser(t *testing.T, fixed map[string]string, history *int) (*Tabnas, map[string]Tin) {
	t.Helper()
	tok := make(map[string]*string, len(fixed))
	for name, src := range fixed {
		s := src
		tok[name] = &s
	}
	opts := Options{
		Rule:  &RuleOptions{Start: "top"},
		Fixed: &FixedOptions{Token: tok},
	}
	if history != nil {
		opts.Rewind = &RewindOptions{History: history}
	}
	j := Make(opts)
	tins := make(map[string]Tin, len(fixed))
	for name := range fixed {
		tins[name] = j.Token(name)
	}
	return j, tins
}

func srcs(toks []*Token) []string {
	out := make([]string, len(toks))
	for i, t := range toks {
		out[i] = t.Src
	}
	return out
}

func TestRewindRecordsConsumed(t *testing.T) {
	j, tn := rewindParser(t, map[string]string{"Ta": "a", "Tb": "b", "Tc": "c"}, nil)
	var recorded []string
	j.Rule("top", func(rs *RuleSpec, _ *Parser) {
		rs.open = []*AltSpec{{S: [][]Tin{{tn["Ta"]}, {tn["Tb"]}, {tn["Tc"]}}}}
		rs.close = []*AltSpec{{S: [][]Tin{{TinZZ}}, A: func(r *Rule, ctx *Context) {
			recorded = srcs(ctx.V)
		}}}
	})
	if _, err := j.Parse("abc"); err != nil {
		t.Fatal(err)
	}
	// a, b, c, then the end sentinel (src "").
	want := []string{"a", "b", "c", ""}
	if strings.Join(recorded, ",") != strings.Join(want, ",") {
		t.Errorf("ctx.V = %v, want %v", recorded, want)
	}
}

func TestRewindReplays(t *testing.T) {
	j, tn := rewindParser(t, map[string]string{"Ta": "a", "Tb": "b", "Tc": "c"}, nil)
	var trace []string
	abc := [][]Tin{{tn["Ta"]}, {tn["Tb"]}, {tn["Tc"]}}
	j.Rule("top", func(rs *RuleSpec, _ *Parser) {
		rs.open = []*AltSpec{{S: abc, P: "again", A: func(r *Rule, ctx *Context) {
			trace = append(trace, "first:"+strings.Join(srcs(ctx.V[len(ctx.V)-3:]), ""))
			if err := ctx.Rewind(0); err != nil {
				t.Errorf("rewind: %v", err)
			}
			trace = append(trace, "after-rewind-v-len:"+itoa(len(ctx.V)))
		}}}
		rs.close = []*AltSpec{{S: [][]Tin{{TinZZ}}}}
	})
	j.Rule("again", func(rs *RuleSpec, _ *Parser) {
		rs.open = []*AltSpec{{S: abc, A: func(r *Rule, ctx *Context) {
			trace = append(trace, "second:"+strings.Join(srcs(ctx.V[len(ctx.V)-3:]), ""))
		}}}
	})
	if _, err := j.Parse("abc"); err != nil {
		t.Fatal(err)
	}
	want := "first:abc|after-rewind-v-len:0|second:abc"
	if strings.Join(trace, "|") != want {
		t.Errorf("trace = %q, want %q", strings.Join(trace, "|"), want)
	}
}

func TestRewindPartial(t *testing.T) {
	j, tn := rewindParser(t, map[string]string{"Ta": "a", "Tb": "b", "Tc": "c", "Td": "d"}, nil)
	var second string
	j.Rule("top", func(rs *RuleSpec, _ *Parser) {
		rs.open = []*AltSpec{{
			S: [][]Tin{{tn["Ta"]}, {tn["Tb"]}, {tn["Tc"]}, {tn["Td"]}},
			P: "tail",
			A: func(r *Rule, ctx *Context) {
				// Mark right after 'a' (absolute index 1); rewind undoes bcd.
				if err := ctx.Rewind(1); err != nil {
					t.Errorf("rewind: %v", err)
				}
			},
		}}
		rs.close = []*AltSpec{{S: [][]Tin{{TinZZ}}}}
	})
	j.Rule("tail", func(rs *RuleSpec, _ *Parser) {
		rs.open = []*AltSpec{{S: [][]Tin{{tn["Tb"]}, {tn["Tc"]}, {tn["Td"]}}, A: func(r *Rule, ctx *Context) {
			second = strings.Join(srcs(ctx.V), "")
		}}}
	})
	if _, err := j.Parse("abcd"); err != nil {
		t.Fatal(err)
	}
	if second != "abcd" {
		t.Errorf("after partial rewind, ctx.V = %q, want abcd", second)
	}
}

func TestRewindNoOp(t *testing.T) {
	j, tn := rewindParser(t, map[string]string{"Ta": "a"}, nil)
	ok := false
	j.Rule("top", func(rs *RuleSpec, _ *Parser) {
		rs.open = []*AltSpec{{S: [][]Tin{{tn["Ta"]}}, A: func(r *Rule, ctx *Context) {
			mark := ctx.Mark()
			if err := ctx.Rewind(mark); err != nil {
				t.Errorf("rewind: %v", err)
			}
			ok = ctx.VAbs == mark
		}}}
		rs.close = []*AltSpec{{S: [][]Tin{{TinZZ}}}}
	})
	if _, err := j.Parse("a"); err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Error("no-op rewind changed VAbs")
	}
}

func TestRewindHistoryCap(t *testing.T) {
	cap4 := 4
	j, tn := rewindParser(t, map[string]string{"Ta": "a"}, &cap4)
	maxSeen := 0
	j.Rule("top", func(rs *RuleSpec, _ *Parser) {
		rs.open = []*AltSpec{{S: [][]Tin{{tn["Ta"]}}, A: func(r *Rule, ctx *Context) {
			if len(ctx.V) > maxSeen {
				maxSeen = len(ctx.V)
			}
		}}}
		rs.close = []*AltSpec{
			{S: [][]Tin{{tn["Ta"]}}, B: 1, R: "top"},
			{S: [][]Tin{{TinZZ}}},
		}
	})
	if _, err := j.Parse("a a a a a a a a a a a a a a a a a a a a"); err != nil {
		t.Fatal(err)
	}
	if maxSeen > 2*4 {
		t.Errorf("ctx.V grew to %d, expected <= %d", maxSeen, 2*4)
	}
	if maxSeen < 4 {
		t.Errorf("ctx.V only reached %d, expected >= 4", maxSeen)
	}
}

func TestRewindPastWindowErrors(t *testing.T) {
	cap2 := 2
	j, tn := rewindParser(t, map[string]string{"Ta": "a"}, &cap2)
	var rewindErr error
	five := [][]Tin{{tn["Ta"]}, {tn["Ta"]}, {tn["Ta"]}, {tn["Ta"]}, {tn["Ta"]}, {tn["Ta"]}}
	j.Rule("top", func(rs *RuleSpec, _ *Parser) {
		rs.open = []*AltSpec{{S: five, A: func(r *Rule, ctx *Context) {
			rewindErr = ctx.Rewind(0)
		}}}
		rs.close = []*AltSpec{{S: [][]Tin{{TinZZ}}}}
	})
	if _, err := j.Parse("a a a a a a"); err != nil {
		t.Fatal(err)
	}
	if rewindErr == nil || !strings.Contains(rewindErr.Error(), "outside the retained history") {
		t.Errorf("expected out-of-window error, got %v", rewindErr)
	}
}

func TestRewindDefaultHistory64(t *testing.T) {
	j, tn := rewindParser(t, map[string]string{"Ta": "a"}, nil)
	finalV := 0
	ten := make([][]Tin, 10)
	for i := range ten {
		ten[i] = []Tin{tn["Ta"]}
	}
	j.Rule("top", func(rs *RuleSpec, _ *Parser) {
		rs.open = []*AltSpec{{S: ten, A: func(r *Rule, ctx *Context) {
			finalV = len(ctx.V)
		}}}
		rs.close = []*AltSpec{{S: [][]Tin{{TinZZ}}}}
	})
	if _, err := j.Parse("a a a a a a a a a a"); err != nil {
		t.Fatal(err)
	}
	// Under the default cap of 64, a 10-token parse retains every token.
	if finalV != 10 {
		t.Errorf("ctx.V = %d, want 10", finalV)
	}
}

func TestRewindUnbounded(t *testing.T) {
	zero := 0 // non-positive history → unbounded
	j, tn := rewindParser(t, map[string]string{"Ta": "a"}, &zero)
	maxV := 0
	j.Rule("top", func(rs *RuleSpec, _ *Parser) {
		rs.open = []*AltSpec{{S: [][]Tin{{tn["Ta"]}}, A: func(r *Rule, ctx *Context) {
			if len(ctx.V) > maxV {
				maxV = len(ctx.V)
			}
		}}}
		rs.close = []*AltSpec{
			{S: [][]Tin{{tn["Ta"]}}, B: 1, R: "top"},
			{S: [][]Tin{{TinZZ}}},
		}
	})
	var sb strings.Builder
	for i := 0; i < 200; i++ {
		if i > 0 {
			sb.WriteByte(' ')
		}
		sb.WriteByte('a')
	}
	if _, err := j.Parse(sb.String()); err != nil {
		t.Fatal(err)
	}
	if maxV < 200 {
		t.Errorf("unbounded history retained only %d, expected >= 200", maxV)
	}
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var b [20]byte
	i := len(b)
	for n > 0 {
		i--
		b[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		b[i] = '-'
	}
	return string(b[i:])
}

File: rs/tests/rewind_test.rs
use std::sync::{Arc, Mutex};
use tabnas::{AltSpec, RuleSpec, Tabnas, Tin, TIN_ZZ};

fn rewind_parser(history: Option<Option<usize>>) -> (Tabnas, [Tin; 3]) {
    let mut parser = Tabnas::new();
    parser
        .grammar_json(
            r##"{"options":{"rule":{"start":"top"},"fixed":{"token":{"#TA":"a","#TB":"b","#TC":"c"}}}}"##,
        )
        .unwrap();
    if let Some(history) = history {
        parser.options.rewind.history = history;
    }
    let tins = [
        parser.options.token("#TA").unwrap(),
        parser.options.token("#TB").unwrap(),
        parser.options.token("#TC").unwrap(),
    ];
    (parser, tins)
}

#[test]
fn context_records_consumed_tokens_and_legacy_top_accessors() {
    let (mut parser, [ta, tb, tc]) = rewind_parser(None);
    let seen = Arc::new(Mutex::new(Vec::new()));
    let seen_action = seen.clone();
    parser.action_with_context("record", move |_rule, context| {
        let mut result = context
            .v
            .iter()
            .map(|token| token.src.clone())
            .collect::<Vec<_>>();
        result.push(context.v1().unwrap().src.clone());
        result.push(context.v2().unwrap().src.clone());
        *seen_action.lock().unwrap() = result;
        Ok(())
    });
    let mut top = RuleSpec::new("top");
    top.open.push(AltSpec {
        s: vec![vec![ta], vec![tb], vec![tc]],
        a: vec!["record".into()],
        ..Default::default()
    });
    top.close.push(AltSpec {
        s: vec![vec![TIN_ZZ]],
        ..Default::default()
    });
    parser.rule(top);

    parser.parse("abc").unwrap();
    assert_eq!(*seen.lock().unwrap(), ["a", "b", "c", "c", "b"]);
}

#[test]
fn context_rewind_replays_consumed_tokens_before_prefetched_lookahead() {
    let (mut parser, [ta, tb, _]) = rewind_parser(None);
    parser.action_with_context("rewind", |_rule, context| context.rewind(0));

    let mut top = RuleSpec::new("top");
    top.open.push(AltSpec {
        s: vec![vec![ta], vec![tb]],
        b: 1,
        a: vec!["rewind".into()],
        p: Some("again".into()),
        ..Default::default()
    });
    top.close.push(AltSpec {
        s: vec![vec![TIN_ZZ]],
        ..Default::default()
    });
    parser.rule(top);

    let replayed = Arc::new(Mutex::new(Vec::new()));
    let replayed_action = replayed.clone();
    parser.action_with_context("record-replay", move |_rule, context| {
        *replayed_action.lock().unwrap() =
            context.v.iter().map(|token| token.src.clone()).collect();
        Ok(())
    });
    let mut again = RuleSpec::new("again");
    again.open.push(AltSpec {
        s: vec![vec![ta], vec![tb]],
        a: vec!["record-replay".into()],
        ..Default::default()
    });
    parser.rule(again);

    parser.parse("ab").unwrap();
    assert_eq!(*replayed.lock().unwrap(), ["a", "b"]);
}

#[test]
fn replayed_tokens_are_reannounced_to_lex_subscribers() {
    let (mut parser, [ta, tb, tc]) = rewind_parser(None);
    parser.action_with_context("rewind", |_rule, context| context.rewind(0));

    let mut top = RuleSpec::new("top");
    top.open.push(AltSpec {
        s: vec![vec![ta], vec![tb], vec![tc]],
        a: vec!["rewind".into()],
        p: Some("again".into()),
        ..Default::default()
    });
    top.close.push(AltSpec {
        s: vec![vec![TIN_ZZ]],
        ..Default::default()
    });
    parser.rule(top);

    let mut again = RuleSpec::new("again");
    again.open.push(AltSpec {
        s: vec![vec![ta], vec![tb], vec![tc]],
        ..Default::default()
    });
    parser.rule(again);

    let seen = Arc::new(Mutex::new(String::new()));
    let seen_subscriber = seen.clone();
    parser.subscribe_tokens(move |token| seen_subscriber.lock().unwrap().push_str(&token.src));
    parser.parse("abc").unwrap();
    assert_eq!(*seen.lock().unwrap(), "abcabc");
}

#[test]
fn bounded_history_rejects_an_evicted_mark() {
    let (mut parser, [ta, _, _]) = rewind_parser(Some(Some(2)));
    parser.action_with_context("rewind-too-far", |_rule, context| context.rewind(0));
    let mut top = RuleSpec::new("top");
    top.open.push(AltSpec {
        s: (0..6).map(|_| vec![ta]).collect(),
        a: vec!["rewind-too-far".into()],
        ..Default::default()
    });
    parser.rule(top);

    let error = parser.parse("a a a a a a").unwrap_err();
    assert_eq!(error.code, "internal");
    assert!(error.detail.contains("outside the retained history"));
}

#[test]
fn serialized_rewind_history_supports_limits_and_unbounded_null() {
    let mut parser = Tabnas::new();
    parser
        .grammar_json(r#"{"options":{"rewind":{"history":4}}}"#)
        .unwrap();
    assert_eq!(parser.options.rewind.history, Some(4));
    parser
        .grammar_json(r#"{"options":{"rewind":{"history":null}}}"#)
        .unwrap();
    assert_eq!(parser.options.rewind.history, None);
}
